### PR TITLE
Parsing of the translate function calls in pages and refactoring of translate function

### PR DESCRIPTION
**Pull request by WoZ**
## Motivation

A lot of code in pages uses javascript data structures like objects and arrays to store the data. Current logic allows to use translations only with `<translate>text</translate>` tag. Text that is stored in data structures isn't processed by translate plugin and plugin only processes JSX tags. I added logic that processes AST's `CallExpression` with name `translate` and adds translations to `i18n/en.json` the same way as `JSXElement` `translate` does. 

Also I refactored `translate` function and added second optional argument `description`. It's much better to delegate the logic of preparing translation key with a description or nor to `translate` function and not to overload babel translate plugin by decision-making. Babel translate plugin just reads `desc` attribute from the `translate` tag and passes it to the `translate` function. So, the logic becomes much more straightforward.

If the `translate` function call is made without passing optional second argument (`description`) the default value is used (`no description given`). The same logic works for `<translate>` tag with/without description.

### Usage.

Add the following require statement to the top of the page file (`pages/en/index.js`):
```js
const translate = require("../../server/translate.js").translate;
```
Use `translate` function `pages/en/index.js`
```jsx
<GridBlock
  align="center"
  contents={[
    {
      content: translate('My pretty new content'),
      title: translate('Translation with description', 'custom description')
    },
  ]}
  layout="threeColumn"
/>
```

Produces `i18n/en.json` with lines
```
// ...
  "pages-strings": {
    "My pretty new content|no description given": "My pretty new content",
    "Translation with description|custom description": "Translation with description",
   // ...
  }
```

`write-translations.js` skips `translate` calls with non-string arguments (not string literals like templates and expressions). The following code will not add keys to `i18n/en.json`

```
const str = 'string';

translate(`String`);
translate('string' + 'string');
translate(str);
```

Generator will output warnings and will suggest to fill strings with hands.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

0. I can add functional test to `build-files.tests.js` but I want to ask where to put `<translate>` tags and `translate` calls. As I see, website folder is used not only for tests but for a building of docusaurus.io and such insertions will add undesired functionality to boilerplate. Am I right?
1. string literals without description - `translate('my pretty new content')` succeeded. It generates
```
"My pretty new content|no description given": "My pretty new content"
```
2. string literals with descriptions - `translate('my pretty new content', 'custom description')` succeeded. It generates
```
"My pretty new content|custom description": "My pretty new content",
```
3. non-string literal as a key - ``translate(`my pretty new content`)`` fails. Output:
```
Skipping "translate" call in /path/Docusaurus/website/pages/en/index.js at line 99 when writing to /i18n/en.json. To enable autogeneration you must pass the string you want to translate as a string literal. Expressions and templates are ignored. You can still add desired strings to /i18n/en.json with your hands.

generate.js triggered...
feed.js triggered...
feed.js triggered...
sitemap.js triggered...
```
4.  string literal as a key and non-string literal as a description - ``translate('my pretty new content', `custom description`)`` fails. Output:
```
Skipping "translate" call in /path/Docusaurus/website/pages/en/index.js at line 99 when writing to /i18n/en.json. To enable autogeneration you must pass description as a string literal. Expressions and templates are ignored. You can still add desired strings to /i18n/en.json with your hands.

generate.js triggered...
feed.js triggered...
feed.js triggered...
sitemap.js triggered...
```
5. more than 2 arguments - `translate('arg1', 'arg2', 'arg3')` fails. Output:
```
Skipping "translate" call in /path/Docusaurus/website/pages/en/index.js at line 99 when writing to /i18n/en.json. To enable autogeneration you must pass only 1 or 2 arguments that must be string literals. Expressions and templates are ignored. You can still add desired strings to /i18n/en.json with your hands.

generate.js triggered...
feed.js triggered...
feed.js triggered...
sitemap.js triggered...
```
## Related PRs

Check the concept first, please. Then I add info to the docs and attach PRs.

### CODE REVIEW

**Comment by WoZ** [comment]
@JoelMarcey  any news?

**Comment by JoelMarcey** [comment]
@WoZ Hey! I did a cursory look of this over the weekend. I plan to spend more time reviewing it this week (probably by Wed). I am going to get another release of Docusaurus out today, so this would probably go in the next one after that -- and releases can be quick one after the other. But I don't want to add major changes to the one I plan to push out today.

**Comment by WoZ** [comment]
@JoelMarcey thank you for response. No problem. I just want to ask you to confirm such strategy. I see a lot of things to do with translations. Any steps that differs to described crowdin integration need to be explained and refactored. I see inconsistency in usage of translated_docs... So I'm ready to help but I want to understand plans and limitations.

**Comment by JoelMarcey** [comment]
@WoZ I like the approach of this. So let's keep moving on it. 👍 

Re: adding a functional test, I definitely think we should do that. I suppose for now `build-files.test.js` is the right place for now. Although, a test specifically for translations might more sense long term.

I am in the middle of adding translations to the docusaurus.io site, which I want to finish in the next couple of weeks. So right now we don't have any `<translate>` tags at all. 

In the meantime, maybe we can create dummy file in `docs` that is not directly accessible on the site (e.g., it won't go in the sidebar), it is just there to test translations for now. Unless you have a better suggestion?

But let's continue with your plan here. I think having the test in this PR would be good, along with docs. What do you think?

**Comment by WoZ** [comment]
@JoelMarcey generally, I think that splitting tests to unit and functional is a good idea. Right now `build-files.test.js` cover our needs, so I can add tests to this file.

By creation of a file in `docs` we will cover only `translated_docs` logic. How do you suggest to test `translate` function call? I can add this test in `pages/en/` and check that test string appears in `en.json`. It may be `index.js` or new unused file just for test purposes. What fo you think?

**Comment by WoZ** [comment]
@JoelMarcey absolutely :thumbsup:

File: lib/write-translations.js [hunk 1]
```diff
@@ -137,6 +146,58 @@ function execute() {
...
+                  );
+
+                  break;
+                }
+                description = prepareTranslationText(
+                  path.node.arguments[1].value
+                );
+              case 1:
+                if (path.node.arguments[0].type !== 'StringLiteral') {
+                  console.warn(
```
Comment (yangshun) [inline lib/write-translations.js]: This `console.warn` seems to be the same as above. Please extract it out into a function.

File: lib/write-translations.js [hunk 2]
```diff
@@ -35,6 +38,14 @@ function writeFileAndCreateFolder(file, content) {
   fs.writeFileSync(file, content);
 }
 
+function prepareTranslationText(text) {
+  if (typeof text !== 'string') {
```
Comment (yangshun) [inline lib/write-translations.js]: I think this `typeof` check is redundant. If `text` wasn't a string, `.trim()` would throw an error. We can do without this check 😄

File: lib/server/translate.js [hunk 3]
```diff
@@ -34,7 +36,10 @@ function doesTranslationExist(str, lang) {
   );
 }
 
-function translate(str) {
+function translate(key, description) {
+  const str = `${key}|${
+    description === undefined ? DEFAULT_DESCRIPTION : description
```
Comment (yangshun) [inline lib/server/translate.js]: How about doing: 

```js
function translate(key, description = DEFAULT_DESCRIPTION) {
  const str = `${key}|${description}`;
  ...
```

instead?

**Comment by WoZ** [comment]
@yangshun hi! I'm on the way!

**Comment by yangshun** [comment]
Hey @WoZ, how is it going? I'm closing this first and we can reopen this when you have the time. If you would not be available to fix this, just let us know and we can take over 😄

**Pull request closed without merging**